File: backend/app/content_extraction/zip_safety.py
"""ZIP bomb guards for OOXML mechanical extraction."""

import zipfile

from app.content_extraction.constants import (
    MAX_OOXML_COMPRESSION_RATIO,
    MAX_OOXML_UNCOMPRESSED_BYTES,
    MAX_OOXML_ZIP_ENTRIES,
)


class UnsafeZipError(Exception):
    pass


def validate_zip_archive(archive: zipfile.ZipFile) -> None:
    entries = archive.infolist()
    if len(entries) > MAX_OOXML_ZIP_ENTRIES:
        raise UnsafeZipError("zip entry count exceeds limit")
    total_uncompressed = 0
    for info in entries:
        if info.is_dir():
            continue
        uncompressed = info.file_size
        compressed = max(info.compress_size, 1)
        if uncompressed / compressed > MAX_OOXML_COMPRESSION_RATIO:
            raise UnsafeZipError("zip compression ratio exceeds limit")
        total_uncompressed += uncompressed
        if total_uncompressed > MAX_OOXML_UNCOMPRESSED_BYTES:
            raise UnsafeZipError("zip uncompressed size exceeds limit")

File: client/lib/local/extraction/zip_safety.dart
import 'package:archive/archive.dart';

import 'extraction_constants.dart';

class UnsafeZipError implements Exception {
  UnsafeZipError(this.message);
  final String message;

  @override
  String toString() => 'UnsafeZipError: $message';
}

void validateZipArchive(Archive archive) {
  final entries = archive.files.where((file) => file.isFile).toList();
  if (entries.length > kMaxOoxmlZipEntries) {
    throw UnsafeZipError('zip entry count exceeds limit');
  }
  var totalUncompressed = 0;
  for (final file in entries) {
    final uncompressed = file.size;
    final compressed = _compressedSize(file);
    if (uncompressed / compressed > kMaxOoxmlCompressionRatio) {
      throw UnsafeZipError('zip compression ratio exceeds limit');
    }
    totalUncompressed += uncompressed;
    if (totalUncompressed > kMaxOoxmlUncompressedBytes) {
      throw UnsafeZipError('zip uncompressed size exceeds limit');
    }
  }
}

int _compressedSize(ArchiveFile file) {
  final raw = file.rawContent;
  if (raw != null) {
    final length = raw.getStream().length;
    if (length > 0) {
      return length;
    }
  }
  return file.size > 0 ? file.size : 1;
}
